app/src/hooks/privy/usePrivyLinkWallet.ts

```diff
@@ -0,0 +1,57 @@
+import { useLinkAccount, usePrivy } from "@privy-io/react-auth"
+import { useRef } from "react"
+import { toast } from "sonner"
+
+import { syncPrivyUser } from "@/db/privy"
+
+import { useUser } from "../db/useUser"
+import { useHandlePrivyErrors } from "../useHandlePrivyErrors"
+
+export const usePrivyLinkWallet = (userId: string) => {
+  const isLinking = useRef(false)
+  const { unlinkWallet } = usePrivy()
+  const { invalidate: invalidateUser } = useUser({ id: userId, enabled: false })
+  const onError = useHandlePrivyErrors()
+
+  const handleUnlinkWallet = (address: string) => {
+    toast.promise(unlinkWallet(address), {
+      loading: "Removing wallet address...",
+      success: (updatedPrivyUser) => {
+        syncPrivyUser(updatedPrivyUser).then(() => {
+          invalidateUser()
+        })
+        return "Wallet address removed successfully"
+      },
+      error: (error) => {
+        return error.message
+      },
+    })
+  }
+
+  const { linkWallet } = useLinkAccount({
+    onSuccess: ({ user: updatedPrivyUser, linkedAccount }) => {
+      if (linkedAccount.type === "wallet" && !isLinking.current) {
+        toast.promise(
+          syncPrivyUser(updatedPrivyUser)
```

Comment:
We shouldn't be calling critical actions within the toast. A couple reasons:

1. We shouldnt tie core platform functionality to UI rendering mechanism which can fail for different reasons
2. Error handling is abstracted and can fail